experimental/QuadFormAndIsom/src/hermitian_miranda_morrison.jl

```diff
@@ -0,0 +1,781 @@
+  # Q is where the determinant of our lifts to good precision will live. So
+  # we just need to create the map from G to Q.
+  Q, mQ = quo(FmodFsharp, I)
+
+  SQ, SQtoQ = snf(Q)
+
+  function dlog(x::Vector)
+    @hassert :ZZLatWithIsom 1 length(x) == length(Fsharpdata)
+    return SQtoQ\(mQ(j\(Fsharplog(x))))
+  end
+
+  imgs = elem_type(SQ)[]
+  # For each of our matrices in gene_herm, we do successive P-adic liftings in
+  # order to approximate an isometry of D^{-1}H^#, up to a certain precision
+  # (given by Theorem 6.25 in BH23). We do this for all the primes we have to
+  # consider up to now, and then map the corresponding determinant adeles inside
+  # Q. Since our matrices were approximate lifts of the generators of G, we can
+  # create the map we wanted from those data.
+  for g in gens(G2)
+    ds = elem_type(E)[]
+    for p in S
+      if !_is_special(H, p)
+        push!(ds, one(E))
+      else
+        lp = prime_decomposition(OE, p)
+        P = lp[1][1]
+        k = valuation(N, p)
+        a = valuation(DEQ, P)
+        e = valuation(DEK, P)
+        g_approx = _approximate_isometry(H, H2, g, P, e, a, k, res)
+        push!(ds, det(g_approx))
+      end
+    end
+    push!(imgs, dlog(ds))
+  end
+
+  GSQ, SQtoGSQ, _ = Oscar._isomorphic_gap_group(SQ)
+  f2 = hom(G2, GSQ, gens(G2), SQtoGSQ.(imgs); check=false)
+  f = compose(GtoG2, f2)
+
+  return f
+end
+
+# We check whether for the prime ideal p E_O(L_p) != F(L_p).
+#
+# There exists some prime ideals p which are elementary divisors for the
+# quotients D^{-1}L^#/L and, according to Kir16b, for which the above quotient
+# can still be trivial. We call special those which do not satisfy this second
+# property: considering them, we can actually reduce the size of the finite
+# abelian group in which we will perform our local approximate determinants
+# computations.
+#
+# This function is an import of a function written by Markus Kirschmer in the
+# Magma package about hermitian lattices.
+
+function _is_special(L::Hecke.HermLat, p::Hecke.NfOrdIdl)
+  OE = base_ring(L)
+  E = nf(OE)
+  lp = prime_decomposition(OE, p)
+  if lp[1][2] != 2 || !Oscar.iseven(rank(L))
+    return false
+  end
+  _, _R, S = jordan_decomposition(L, p)
+  R = [nrows(m) for m in _R]
+  for r in R
+    if r != 2
+      return false
+    end
+  end
+  P = lp[1][1]
+  e = valuation(different(OE), P)
+  for i in 1:length(S)
+    if !iseven(e- S[i])
+      return false
+    end
+  end
+  u = elem_in_nf(uniformizer(P))
+  s = involution(L)
+  su = s(u)
+  H = block_diagonal_matrix([matrix(E, 2, 2, [0 u^(S[i]); su^(S[i]) 0]) for i in 1:length(S)])
+  return is_locally_isometric(L, hermitian_lattice(E, gram = H), p)
+end
+
+###############################################################################
+#
+#  Local hermitian lifting - path to algorithm 8 of BH23
+#
+###############################################################################
+
+Oscar.canonical_unit(x::NfOrdQuoRingElem) = one(parent(x))
```

Comment:
```suggestion
Oscar.canonical_unit(x::NfOrdQuoRingElem) = one(parent(x))
```
Why?

Reply:
I do not remember, but one of the few algorithms I was using was buggy because it needed a canonical unit for such kind of objects. Then I moved stuff around, so I do not know whether I still need it, but at some point I did

Reply:
Maybe for `solve_left`, now that I think about it. It looks needed whenever we do some kind of Gauss elimination (cf line 582)